### ISSUE TITLE
[FEATURE REQUEST] gio::File `Eq` implementation does not use `equal`

### ISSUE THREAD

**Issue by euclio** [gtk-rs/gtk-rs-core#952]
I had a bug in my code because the `gio::File` `Eq` implementation compares pointer values instead of delegating to `g_file_equal`.

This behavior is unintuitive, and ideally should be fixed, or at least documented on the `Eq` implementation.

**Comment by sdroege** [comment]
While this sucks indeed, the alternative would be to make the `Eq` / `PartialEq` impls incoherent: Only `gio::File` (and other selected types) have a proper equality function, but if you compare a `gio::File` (as `glib::Object`) to another object it would still do just the pointer check because there's no virtual method for equality in `glib::Object`.

See https://gitlab.gnome.org/GNOME/glib/-/issues/1671 for context and a similar issue around the `Debug` impl.

**Comment by jf2048** [comment]
Do any other gi bindings provide overloads for equality? To be more correct with Rust would be if we had [`ptr_eq`](https://doc.rust-lang.org/stable/std/sync/struct.Arc.html#method.ptr_eq) for objects anyway

**Comment by sdroege** [comment]
> Do any other gi bindings provide overloads for equality

How would that work if GObject doesn't provide a way to overload equality?

> To be more consistent with `std` would be if we had [`ptr_eq`](https://doc.rust-lang.org/stable/std/sync/struct.Arc.html#method.ptr_eq) for objects anyway, i.e. in Rust the `==` operator is for value comparisons so is reversed from how it is in C# and Java

Right now (until GObject has support for a comparison vfunc), there's only one possible way to implement object equality and that's the pointer equality we have right now. That seems less of a problem than the alternative to not have any `PartialEq`, `Eq`, `PartialOrd`, `Ord` and `Hash` impls at all.

**Comment by jf2048** [comment]
I mean manually implemented ones like we would have to do. AFAIK there are no gobjects that really have it but there are some boxed/shared types like `GDateTime`, `GstCaps`, `GtkBitset`, etc that provide equality functions

**Comment by sdroege** [comment]
Yes but those don't have the complications of inheritance so that's easy.
